doc/DESIGN.md

```diff
@@ -283,3 +283,107 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
 The reason behind this is, that the above functions – as their name
 suggest – only retrieve values. The returned value must not be modified
 directly.
+
+## Design Guidelines Checklist
+
+On potential changes of the API/ABI as detected by
+[`elektra-icheck`](https://build.libelektra.org/job/elektra-icheck),
+please make sure the API has been reviewed according to the
+following 2 checklists:
+
+## Checklist for overall API
+
+### Naming
+
+- [ ] Inconsistent naming of functions or variables
+
+### Consistency
+
+- [ ] Consistent naming schemes
+- [ ] Similar things are named similarly
+- [ ] Different things are named differently
+- [ ] The order of arguments should be consistent across similar functions
+- [ ] Code is formatted according to the [style guidelines](/doc/CODING.md#coding-style)
+
+### Structural Clarity
+
+- [ ] Functions with similar functionality are grouped into the same namespace
+
+### Compatibility
+
+- [ ] All Bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
+- [ ] Components do not depend too heavily on each other
+
+## Checklist for each function
+
+### Documentation
+
+- [ ] Doxygen Documentation is complete
+      (covers all parameters, brief/short summary, examples)
+- [ ] Change is mentioned in the release notes
+- [ ] Inconsistencies between documentation and code
+- [ ] Inconsistencies between documentation and tests
+- [ ] Proper Documentation of all side effects
+- [ ] Proper Documentation of thread-safety of function
+- [ ] [Symbol versioning](/doc/dev/symbol-versioning.md)
+      is correct for breaking changes
+
+### Naming
+
+- [ ] Abbreviations should be avoided in function / parameter names
+- [ ] Function / parameter names should neither be too long, nor too short
+- [ ] Function name should be clear and unambiguous
+
+### Compatibility
+
+- [ ] ABI/API forward-compatible (breaking backwards-compatibility
+      to add new symbols is fine)
+- [ ] #ifdef present for experimental features
+
+### Parameter & Return Types
+
+- [ ] Returning a not specific enough type
+- [ ] Requiring a not liberal enough parameter
+- [ ] should not have multiple parameters of the same type
+- [ ] Functions should use Constant Types instead of Boolean Types wherever
+      sensible
+- [ ] Wherever possible, function parameters should be `const`
+- [ ] Functions should not have a long list of parameters (>8)
```

Comment:
At the moment we do not even have 4.

Reply:
Nice, should I reduce the number included or remove the whole line altogether? This was more or less arbitrarily chosen and can definitely be change to accomodate the project.

Reply:
This is basically the counterpart of being generic. With more generic functions (e.g. flags to modify behavior) more parameters are added. Maybe we simply add here in the checklist that these trade-offs should be considered, without giving numbers or judgements how the trade-off should be chosen as this is already described in doc/GOALS.md